File: 713. Subarray Product Less Than K/solution.py
class Solution:
    def numSubarrayProductLessThanK(self, nums, k: int) -> int:
        # Edge case: If k is 0 or 1, no product of positive integers can be < k
        if k <= 1:
            return 0

        l = 0              # Left pointer of the sliding window
        product = 1        # Current product of the window
        count = 0          # Total number of valid subarrays

        # Iterate over each element using the right pointer
        for r in range(len(nums)):
            product *= nums[r]  # Expand the window by multiplying the current element

            # While the product is greater than or equal to k, shrink the window from the left
            while product >= k:
                product /= nums[l]
                l += 1

            # For each valid window, the number of subarrays ending at index r is (r - l + 1)
            # This is because all subarrays from l to r, (l+1) to r, ..., r to r are valid
            count += (r - l + 1)

        return count

# Time Complexity: O(n) — Each element is visited at most twice (once by r, once by l)
# Space Complexity: O(1) — Uses constant extra space

File: 713. Subarray Product Less Than K/solution.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        // Edge case: If k is 0 or 1, no product of positive integers can be less than k
        if (k <= 1) {
            return 0;
        }

        int l = 0;         // Left pointer of the sliding window
        int product = 1;   // Current product of elements in the window
        int count = 0;     // Total count of valid subarrays

        // Iterate through the array with the right pointer
        for (int r = 0; r < nums.size(); r++) {
            product *= nums[r];  // Include nums[r] in the product

            // Shrink the window from the left until the product is less than k
            while (product >= k) {
                product /= nums[l];
                l++;
            }

            // For each valid window, the number of subarrays ending at index r is (r - l + 1)
            count += (r - l + 1);
        }

        return count;
    }
};
